### PR TITLE
Add custom parameter for a sensu_check

### PR DESCRIPTION
**Pull request by mbobakov**

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @mbobakov. @andsens please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

File: monitoring/sensu_check.py [hunk 1]
```diff
@@ -149,6 +149,11 @@
       - The low threshhold for flap detection
     required: false
     default: null
+  custom:
+    description:
+      - JSON mixin to add to the configuration
```
Comment (andsens) [inline monitoring/sensu_check.py]: You should specify whether other existing keys are overwritten (you throw an error, I know, but the user should be able to know this by reading the docs).

File: monitoring/sensu_check.py [hunk 2]
```diff
@@ -183,6 +188,7 @@
         # Let snippet from module_utils/basic.py return a proper error in this case
         pass
 
+import ast
```
Comment (andsens) [inline monitoring/sensu_check.py]: ~~Why not use the json module that was imported 4 lines prior?~~

Ignore this, just use the `dict` type for the ansible param.

File: monitoring/sensu_check.py [hunk 3]
```diff
@@ -257,6 +263,36 @@ def sensu_check(module, path, name, state='present', backup=False):
                     changed = True
                     reasons.append('`{opt}\' was removed'.format(opt=opt))
 
+        if module.params['custom']:
+          # Convert to json
+          try:
+            custom_params = ast.literal_eval(module.params['custom'])
```
Comment (andsens) [inline monitoring/sensu_check.py]: ~~The python `dict` notation is not the same as the JSON notation. You should really use the JSON module here.~~

Ignore this, just use the `dict` type for the ansible param.

File: monitoring/sensu_check.py [hunk 4]
```diff
@@ -257,6 +263,36 @@ def sensu_check(module, path, name, state='present', backup=False):
...
+          except:
+            msg = 'Module parameter "custom" contains invalid JSON. Example: custom=\'{"JSON": "here"}\''
+            module.fail_json(msg=msg)
+
+          overwrited_fields = set(custom_params.keys()) & set(simple_opts + ['type','subdue'])
```
Comment (andsens) [inline monitoring/sensu_check.py]: You're missing `subdue_end` and `subdue_begin` (`subdue` is the name of the param in sensu and is an object with `begin` and `end`, in ansible we simply split the two subelements up into two options)

File: monitoring/sensu_check.py [hunk 5]
```diff
@@ -257,6 +263,36 @@ def sensu_check(module, path, name, state='present', backup=False):
...
+          if overwrited_fields:
+            msg = 'You can\'t overwriting standard module parameters via "custom". You are trying overwrite: {of}'.format(of=list(overwrited_fields))
+            module.fail_json(msg=msg)
+
+          for k,v in custom_params.items():
```
Comment (andsens) [inline monitoring/sensu_check.py:274]: Use `opt, value` like above, for consistency.

File: monitoring/sensu_check.py [hunk 6]
```diff
@@ -257,6 +263,36 @@ def sensu_check(module, path, name, state='present', backup=False):
...
+            if k in config['checks'][name].keys():
```
Comment (andsens) [inline monitoring/sensu_check.py]: The `in` operator checks the keys when given a dict, so there\s no need for `.keys()`

File: monitoring/sensu_check.py [hunk 7]
```diff
@@ -320,6 +356,7 @@ def main():
                 'aggregate':    {'type': 'bool'},
                 'low_flap_threshold':  {'type': 'int'},
                 'high_flap_threshold': {'type': 'int'},
+                'custom':   {'type': 'str'},
```
Comment (andsens) [inline monitoring/sensu_check.py]: You can just use `dict` as the type here. In fact, that would make the code a lot simpler. Ansible will then make sure that your param is properly formatted :-)

**Comment by andsens** [comment]
Great job @mbobakov! Kudos for following the style of the existing code. I added a few comments. If you could fix those, we can get this merged :-)

needs_revision

File: monitoring/sensu_check.py [hunk 8]
```diff
@@ -257,6 +263,36 @@ def sensu_check(module, path, name, state='present', backup=False):
...
+                changed = True
+                reasons.append('`custom param {k}\' was changed'.format(k=k))
+            else:
+              changed = True
+              reasons.append('`custom param {k}\' was added'.format(k=k))
+            check[k] = v
+          simple_opts += custom_params.keys()
+
+        # Remove obsolete custom params
+        for opt in set(config['checks'][name].keys()) - set(simple_opts + ['type','subdue']):
```
Comment (andsens) [inline monitoring/sensu_check.py]: Don't forget `subdue_begin` and `subdue_end` here as well.

**Comment by mbobakov** [comment]
Hello.
Dictionary, exactly! Thank you for a tip. I just did not think about it.
I have fixed a code, can you check it again?

**Comment by andsens** [comment]
Ah :-) That made the code a lot simpler. 
Looks great!

shipit

**Comment by gregdek** [comment]
Thanks again to @mbobakov for this PR, and thanks @andsens for reviewing. Marking for inclusion.

**Pull request merged**